floors/rule.go

```diff
@@ -0,0 +1,381 @@
+package floors
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/bits"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/buger/jsonparser"
+	"github.com/mxmCherry/openrtb/v16/openrtb2"
+	"github.com/prebid/prebid-server/currency"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+const (
+	SiteDomain string = "siteDomain"
+	PubDomain  string = "pubDomain"
+	Domain     string = "domain"
+	Bundle     string = "bundle"
+	Channel    string = "channel"
+	MediaType  string = "mediaType"
+	Size       string = "size"
+	GptSlot    string = "gptSlot"
+	PbAdSlot   string = "pbAdSlot"
+	Country    string = "country"
+	DeviceType string = "deviceType"
+	Tablet     string = "tablet"
+	Phone      string = "phone"
+)
+
+func getFloorCurrency(floorExt *openrtb_ext.PriceFloorRules) string {
+	floorCur := "USD"
+	if floorExt.Data.Currency != "" {
+		floorCur = floorExt.Data.Currency
+	}
+
+	if floorExt.Data.ModelGroups[0].Currency != "" {
+		floorCur = floorExt.Data.ModelGroups[0].Currency
+	}
+	return floorCur
+}
+
+func getMinFloorValue(floorExt *openrtb_ext.PriceFloorRules, conversions currency.Conversions) (float64, string, error) {
+	var err error
+	var rate float64
+	floorMin := floorExt.FloorMin
+	floorCur := getFloorCurrency(floorExt)
+
+	if floorExt.FloorMin > 0.0 && floorExt.FloorMinCur != "" && floorCur != "" &&
+		floorExt.FloorMinCur != floorCur {
+		rate, err = conversions.GetRate(floorExt.FloorMinCur, floorCur)
+		floorMin = rate * floorExt.FloorMin
+	}
+	return floorMin, floorCur, err
+}
+
+func updateImpExtWithFloorDetails(matchedRule string, imp *openrtb2.Imp, floorVal float64) {
+	imp.Ext, _ = jsonparser.Set(imp.Ext, []byte(`"`+matchedRule+`"`), "prebid", "floors", "floorRule")
+	imp.Ext, _ = jsonparser.Set(imp.Ext, []byte(fmt.Sprintf("%.4f", floorVal)), "prebid", "floors", "floorRuleValue")
```

Comment:
You may not use `jsonparser.Set`. it is not production ready code and has caused panics for us in the past.

Reply:
You need to change this to use the standard Go json library. We've tried to use `jsonparser.Set`before and it seg faulted on us in certain cases.